File: src/child.rs
/*! Copyright (C) 2022 Muqiu Han
 **
 ** This program is free software: you can redistribute it and/or modify
 ** it under the terms of the GNU General Public License as published by
 ** the Free Software Foundation, either version 3 of the License, or
 ** (at your option) any later version.
 **
 ** This program is distributed in the hope that it will be useful,
 ** but WITHOUT ANY WARRANTY; without even the implied warranty of
 ** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 ** GNU General Public License for more details.
 **
 ** You should have received a copy of the GNU General Public License
 ** along with this program.  If not, see <https://www.gnu.org/licenses/>. */

use std::path::PathBuf;

use crate::config::ContainerConfig;
use crate::error::ErrorCode;

use crate::hostname::set_container_hostname;
use crate::mount;
use crate::mount::set_mount_point;
use crate::syscall::setsyscalls;
use nix::mount::MsFlags;
use nix::sched::clone;
use nix::sched::CloneFlags;
use nix::sys::signal::Signal;
use nix::unistd::execve;
use nix::unistd::Pid;
use std::ffi::CString;

const STACK_SIZE: usize = 1024 * 1024;

fn child(config: ContainerConfig) -> isize {
    info!(
        "starting container with command {} and args {:?}",
        config.path.to_str().unwrap(),
        config.argv
    );

    let retcode = match execve::<CString, CString>(&config.path, &config.argv, &[]) {
        Ok(_) => 0,
        Err(e) => {
            error!("error while trying to perform execve: {:?}", e);
            -1
        }
    };

    match setup_container_configurations(&config) {
        Ok(_) => info!("container setup successfully"),
        Err(e) => {
            error!("error while setup container: {:?}", e);
            return -1;
        }
    }

    retcode
}

fn setup_container_configurations(config: &ContainerConfig) -> Result<(), ErrorCode> {
    set_container_hostname(&config.hostname)?;
    set_mount_point(&config.mount_dir, &config.add_paths)?;

    mount::wrapper::mount_dir(
        None,
        &PathBuf::from("/"),
        vec![MsFlags::MS_REC, MsFlags::MS_PRIVATE],
    )?;

    setsyscalls()?;
    Ok(())
}

fn child_flags() -> CloneFlags {
    let mut flags = CloneFlags::empty();

    // Start the cloned subroutine in a new mounted namespace, initialized with a copy of the parent process's namespace.
    flags.insert(CloneFlags::CLONE_NEWNS);

    // Start the cloned child process in a new cgroup namespace.
    flags.insert(CloneFlags::CLONE_NEWCGROUP);

    // Start the cloned child process in a new pid namespace.
    flags.insert(CloneFlags::CLONE_NEWPID);

    // Start the cloned child process in a new ipc namespace.
    flags.insert(CloneFlags::CLONE_NEWIPC);

    // Start the cloned subroutine in a new network namespace.
    flags.insert(CloneFlags::CLONE_NEWNET);

    // Start the cloned child process in a new uts namespace.
    flags.insert(CloneFlags::CLONE_NEWUTS);

    flags
}

pub fn generate_child_process(config: ContainerConfig) -> Result<Pid, ErrorCode> {
    let mut tmp_stack: [u8; STACK_SIZE] = [0; STACK_SIZE];

    match clone(
        Box::new(|| child(config.clone())),
        &mut tmp_stack,
        child_flags(),
        Some(Signal::SIGCHLD as i32),
    ) {
        Ok(pid) => Ok(pid),
        Err(_) => Err(ErrorCode::ChildProcessError(0)),
    }
}

File: src/child.cpp
/** Copyright (C) 2023 Muqiu Han <muqiu-han@outlook.com> */

#include "include/child.h"
#include "include/capabilities.h"
#include "include/exec.h"
#include "include/hostname.h"
#include "include/log.hpp"
#include "include/mount.h"
#include "include/namespace.h"
#include "include/syscall.h"

#include <__expected/expected.h>
#include <cstdio>
#include <error.h>
#include <sched.h>
#include <sys/wait.h>

namespace bonding::child
{
  std::expected<void, error::Err> Child::Process::setup_container_configurations() noexcept
  {
    hostname::Hostname::setup(container_options->hostname).value();
    mounts::Mount::setup(
      container_options->mount_dir, container_options->hostname, container_options->mounts)
      .value();
    ns::Namespace::setup(container_options->ipc.second, container_options->uid).value();
    capabilities::Capabilities::setup().value();
    syscall::Syscall::setup().value();

    return {};
  }

  int Child::Process::_main(void *options) noexcept
  {
    container_options = static_cast<config::Container_Options *>(options);

    setup_container_configurations()
      .transform([]() -> std::expected<void, error::Err> {
      LOG_INFO << "Container setup successfully";
      return {};
    })
      .transform_error([&](const error::Err &err) {
      LOG_ERROR << "Error while creating container";
      return err;
    }).value();

    int ret_code = 0;

    if (!exec::Execve::call(container_options->path, container_options->argv).has_value())
      ret_code = -1;

    return ret_code;
  }

  std::expected<pid_t, error::Err> Child::generate_child_process(
    const config::Container_Options container_options) noexcept
  {
    const pid_t child_pid = clone(
      Process::_main,
      static_cast<char *>(Process::STACK) + Process::STACK_SIZE,
      container_options.clone_flags,
      (void *) &container_options);

    if (-1 == child_pid)
      return std::unexpected(ERR(error::Code::ChildProcess));

    return child_pid;
  }

  std::expected<void, error::Err> Child::wait() const noexcept
  {
    LOG_DEBUG << "Waiting for child process " << m_pid << " finish...";

    int child_process_status = 0;

    /** To wait for children produced by clone(), need __WCLONE flag */
    if (-1 == waitpid(m_pid, &child_process_status, __WALL))
      return std::unexpected(ERR(error::Code::Container));

    LOG_INFO << "Child process exit with code " << ((child_process_status >> 8) & 0xFF)
             << ", signal " << (child_process_status & 0x7F);

    return {};
  }
} // namespace bonding::child
